### PR TITLE
Add ChecksumAlgorithm to csc/vbc build task

### PR DESCRIPTION
**Pull request by agocke**
/cc @tmat @jaredpar @jasonmalinowski @dotnet/roslyn-compiler

### CODE REVIEW

**Comment by agocke** [comment]
AskMode motivation: this is necessary for partners who have restrictions around using SHA1 for any purpose, even non-security-critical purposes.

**Comment by agocke** [comment]
+ @MattGertz for ask mode approval, pending PR approval

File: src/Compilers/Core/MSBuildTaskTests/CscTests.cs [hunk 1]
```diff
@@ -91,6 +91,20 @@ public void RuntimeMetadataVersionFlag()
...
+        public void ChecksumAlgorithmFlag()
+        {
+            var csc = new Csc();
+            csc.Sources = MSBuildUtil.CreateTaskItems("test.cs");
+            csc.ChecksumAlgorithm = "sha256";
+            Assert.Equal("/out:test.exe /checksumalgorithm:sha256 test.cs", csc.GenerateResponseFileContents());
+
+            csc = new Csc();
+            csc.Sources = MSBuildUtil.CreateTaskItems("test.cs");
+            csc.ChecksumAlgorithm = null;
```
Comment (tmat) [inline src/Compilers/Core/MSBuildTaskTests/CscTests.cs]: What about an empty string?

**Comment by jasonmalinowski** [comment]
From the IDE perspective, this is emit only? Does it have any impact on edit and continue?

File: src/Compilers/Core/MSBuildTaskTests/CscTests.cs (hunk 1 above)
Comment (agocke) [inline src/Compilers/Core/MSBuildTaskTests/CscTests.cs]: I'll add a test, but this is not representable in MSBuild. Passing an empty value to the property of  a task is the same as assigning its default value (null).

**Comment by agocke** [comment]
@jasonmalinowski This is emit only. It could have an effect on edit-and-continue if a new PDB is emitted, since theoretically that PDB should also use the specified checksum algorithm. What kind of effect are you trying to protect against?

File: src/Compilers/Core/MSBuildTaskTests/CscTests.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/Compilers/Core/MSBuildTaskTests/CscTests.cs]: Never underestimate the number of places people manually create tasks.

**Comment by jasonmalinowski** [comment]
Emit only options aren't consumed by Edit and Continue as far as I know, so this would mean that anybody who sets this will get mismatches from the command line compiler. Is that bad?

**Comment by agocke** [comment]
@jasonmalinowski Ah, sorry, I interpreted your question as whether it only affected the emit output of the compiler, not whether it was part of the `EmitOptions`. It is not, it's included in the `CommandLineArguments` and then encoded into the `SourceText` for the parsed files. What method is used by the IDE to parse these files?

**Comment by agocke** [comment]
@tmat Is there any situation in which EnC could cause a new PDB to be emitted to the output directory with the wrong checksum? If there's no way the EnC PDB could be used in place of the "compiled" PDB there shouldn't be anything to worry about.

**Comment by jasonmalinowski** [comment]
@agocke: let's test Edit and Continue with this flag on just to make sure nothing bizarre happens. I agree nothing should, but let's test either way.

**Comment by tmat** [comment]
EnC should be fine. Each change is compiled to a new PDB that only contains methods affected by the change. The PDB is not saved anywhere. It would certainly be good to test this but other than manual test or integration tests there is no easy way to do so in a unit test, since the document hash is only consumed by the VS debugger, which is not unit-testable.

**Comment by jaredpar** [comment]
👍

**Comment by agocke** [comment]
@tmat @jasonmalinowski I manually tested EnC and verified the PDB's output with a hex editor. Everything looks fine.

**Comment by agocke** [comment]
@tmat @jasonmalinowski Ping

**Comment by tmat** [comment]
:+1:

**Comment by jasonmalinowski** [comment]
:+1:

**Pull request merged**